server/repository/src/db_diesel/temperature_chart/temperature_chart_row.rs

```diff
@@ -0,0 +1,224 @@
+use crate::{db_diesel::temperature_log_row::temperature_log, DBType};
+use chrono::NaiveDateTime;
+use diesel::{
+    backend::Backend, expression::*, prelude::*, query_builder::*, query_source::*, sql_types::*,
+};
+
+#[derive(Debug, PartialEq)]
+pub struct TemperatureChartRow {
+    pub from_datetime: NaiveDateTime,
+    pub to_datetime: NaiveDateTime,
+    pub average_temperature: f64,
+    pub sensor_id: String,
+}
+
+// Needed for allow_table_to_appear_in_same_query
+
+pub use TemperatureChart as table;
+allow_tables_to_appear_in_same_query!(temperature_log, self);
+
+#[derive(Debug, Clone, QueryId)]
+pub struct TemperatureChart {
+    pub(super) intervals: Vec<Interval>,
+}
+#[derive(Debug, Clone, PartialEq)]
+pub struct Interval {
+    pub from_datetime: NaiveDateTime,
+    pub to_datetime: NaiveDateTime,
+}
+
+impl QueryFragment<DBType> for TemperatureChart {
+    fn walk_ast(&self, mut out: AstPass<DBType>) -> QueryResult<()> {
+        // Below should produces something like
+        // (
+        //  SELECT '2021-01-01T16:00:00' as from_datetime, '2021-01-01T17:00:00' as to_datetime
+        //  UNION SELECT '2021-01-01T17:00:00' as from_datetime, '2021-01-01T18:00:00' as to_datetime
+        //  UNION SELECT '2021-01-01T18:00:00' as from_datetime, '2021-01-01T19:00:00' as to_datetime
+        //  ) AS time_series
+        //  JOIN temperature_log ON
+        //       (temperature_log.datetime >= time_series.from_datetime
+        //       AND temperature_log.datetime < time_series.to_datetime)
+
+        let mut union_prefix = "";
+
+        out.push_sql(" ( ");
+        for Interval {
+            from_datetime: from_date,
+            to_datetime: to_date,
+        } in self.intervals.iter()
+        {
+            out.push_sql(union_prefix);
+            // Only add UNION after first select
+            union_prefix = " UNION ";
+
+            out.push_sql(" SELECT ");
+            out.push_bind_param::<Timestamp, _>(&from_date)?;
+            out.push_sql(" as from_datetime, ");
+            out.push_bind_param::<Timestamp, _>(&to_date)?;
+            out.push_sql(" as to_datetime ");
+        }
+
+        out.push_sql(
+            r#" 
+                ) AS time_series
+               JOIN temperature_log ON 
+                    (temperature_log.datetime >= time_series.from_datetime
+                    AND temperature_log.datetime < time_series.to_datetime)"#,
+        );
+
+        Ok(())
+    }
+}
+
+impl QuerySource for TemperatureChart {
+    type FromClause = Self;
+    type DefaultSelection = AllColumns;
+    fn from_clause(&self) -> Self::FromClause {
+        // In expanded macro this Identifier("table_name"), which is translates to `table_name`
+        // in walk_ast. We change this to wal_ast of TemperatureLog
+        self.clone()
```

Comment:
This is the main change from expanded macro (logical change, since it's quite a lot more simplified syntax wise)